docs/subprocess-serialization.md

```diff
@@ -0,0 +1,198 @@
+# Tensorizer serialization via subprocess
+
+If you're using Tensorizer serialization to write checkpoints during training,
+you may want to run the serialization concurrently from your training code so
+that you can execute your next training step as quickly as possible.  And
+because of the Python GIL, it's better to do this in a separate process so that the
+serialization doesn't utilize any of the GIL that you'd otherwise use in your training code.
+
+Keep in mind that this is a way to achieve _concurrency_, not instant
+snapshotting. The tensors you are checkpointing still need to be kept in memory,
+unmodified, for the duration of the serialization process. (Though you may
+choose to copy them out of CUDA memory into CPU memory. These tradeoffs are
+discussed below.)
+
+Also refer to [PyTorch Multiprocessing best
+practices](https://pytorch.org/docs/stable/notes/multiprocessing.html) for more
+details about using PyTorch across processes
+
+
+## Warning about fork() and threads
+Be aware that Python `os.fork()` is often not a viable option, as it can be known to cause deadlocks if you have multiple threads. Python 3.12 and above
+will [issue a deprecation warning](https://github.com/python/cpython/pull/100229) if you attempt this. Some 3rd-party packages that rely on sockets or file descriptors may also not behave correctly when a process unexpectedly forks. 
+
+A subprocess (fork + exec) is generally safer, but you do not inherently get
+shared memory with the calling process. `multiprocessing` has two ways to create
+a child process: `spawn` or `forkserver`. `spawn` should always be safe.
+`forkserver` can be faster but safety depends on the behavior of modules at
+import time. See
+https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods
+for more details.
+
+## If starting from CUDA
+Presuming your tensors are in CUDA memory, there are a couple different options.
+
+### Option 1: Communicate the CUDA tensors directly
+CUDA tensors can be "shared" to a subprocess very efficiently since it's only communicating a pointer to device memory.
+
+Basically send the CUDA tensors over a `multiprocessing.Queue` to a subprocess that does the serialization. Ensure that the CUDA tensors remain **unmodified** in device memory until the serialization process finishes. Note if using encryption, the tensors may be encrypted-in-place so they must also not read again until serialization finishes.
```

Comment:
Wait I forgot, the prohibition on reading when using encryption only applies for CPU tensors, since when serializing from CUDA it only modifies the temporary CPU copy. We can drop that sentence. My bad.
```suggestion
Basically send the CUDA tensors over a `multiprocessing.Queue` to a subprocess that does the serialization. Ensure that the CUDA tensors remain **unmodified** in device memory until the serialization process finishes.
```

Reply:
lol jinx